### PR TITLE
Make PackageDownloadsDetail and PackageDownloadsByVersion graphs update without refreshing the page

### PR DESCRIPTION
**Pull request by scottbommarito**
This is an accessibility fix that has developed into a user experience improvement in general.

Previously, the [PackageDownloadsDetail](https://www.nuget.org/stats/packages/newtonsoft.json/10.0.1) and [PackageDownloadsByVersion](https://www.nuget.org/stats/packages/newtonsoft.json) pages would refresh every time the selected dimensions were changed. This was obviously bad for accessibility, as the focus would be lost every time the page refreshed, which disorients users.

I have fixed this.

- Previously, the logic to inflate the table of information used by the browser was done in CSHTML, which is created server side. This logic is now done in JS using knockout (see `_PivotTable.cshtml`). We already had knockout in our project so I used the existing version.

- Added a route that returns the statistics report in JSON.

- Clicking upon a checkbox now makes an AJAX call that, on success, reinflates the information table using the statistics report in JSON. On error, it shows a message that asks the user if they would like to try again, which can be clicked to retry the AJAX call.

Tested on all the major browsers (`Firefox`, `Chrome`, `Edge`).

### CODE REVIEW

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 1]
```diff
@@ -158,6 +158,63 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
...
+
+        public virtual async Task<JsonResult> PackageDownloadsByVersionReport(string id, string[] groupby)
+        {
+            return Json(await GetPackageDownloadsByVersionReport(id, groupby), JsonRequestBehavior.AllowGet);
+        }
+
+        //
+        // GET: /stats/packages/{id}/{version}
+
+        public virtual async Task<ActionResult> PackageDownloadsDetail(string id, string version, string[] groupby)
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: please add UTs for the new endpoints.

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 2]
```diff
@@ -158,6 +158,63 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
...
+
+            model.UseD3 = UseD3();
+
+            return View(model);
+        }
+
+        //
+        // GET: /stats/reports/packages/{id}
+
+        public virtual async Task<JsonResult> PackageDownloadsByVersionReport(string id, string[] groupby)
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: consider adding basic functional tests to check those endpoints

Comment (scottbommarito) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: Existing functional tests have been adapted to check these new endpoints.

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 3]
```diff
@@ -158,6 +158,63 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
...
+
+            return View(model);
+        }
+
+        //
+        // GET: /stats/reports/packages/{id}/{version}
+
+        public virtual async Task<JsonResult> PackageDownloadsDetailReport(string id, string version, string[] groupby)
+        {
+            var report = await GetPackageDownloadsDetailReport(id, version, groupby);
```
Comment (chenriksson) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: Should the new methods also check if (_statisticsService == NullStatisticsService.Instance)?

File: src/NuGetGallery/Scripts/statsdimensions.js [hunk 4]
```diff
@@ -1,5 +1,49 @@
...
+        url: baseUrl + query,
+        type: 'GET',
+        dataType: 'json',
+        success: renderGraphHandler,
+        error: function () {
+            renderGraphHandler(null);
+
+            $('#statistics-retry').click(function () {
+                renderGraph(baseUrl, query);
+            });
```
Comment (chenriksson) [inline src/NuGetGallery/Scripts/statsdimensions.js]: Does it need to register a new click callback on every failure?

Comment (scottbommarito) [inline src/NuGetGallery/Scripts/statsdimensions.js]: Yes, `query` can change.

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 5]
```diff
@@ -158,6 +158,63 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
                 return new HttpStatusCodeResult(HttpStatusCode.NotFound);
             }
 
+            var report = await GetPackageDownloadsByVersionReport(id, groupby);
+
+            StatisticsPackagesViewModel model = new StatisticsPackagesViewModel();
```
Comment (loic-sharma) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: var?

Comment (scottbommarito) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: was existing code, but i'll fix

File: src/NuGetGallery/Controllers/StatisticsController.cs (hunk 3 above)
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: will add

File: src/NuGetGallery/Controllers/StatisticsController.cs (hunk 1 above)
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: done

**Pull request merged**